Code/Tools/FBuild/FBuildCore/Graph/DependencyListNode.cpp

```diff
@@ -0,0 +1,138 @@
+// DependencyListNode.cpp
+//------------------------------------------------------------------------------
+
+// Includes
+//------------------------------------------------------------------------------
+#include "DependencyListNode.h"
+
+#include "Tools/FBuild/FBuildCore/FBuild.h"
+#include "Tools/FBuild/FBuildCore/FLog.h"
+#include "Tools/FBuild/FBuildCore/BFF/Functions/Function.h"
+#include "Tools/FBuild/FBuildCore/Graph/NodeGraph.h"
+
+#include "Core/Containers/Array.h"
+#include "Core/Env/ErrorFormat.h"
+#include "Core/FileIO/FileIO.h"
+#include "Core/FileIO/FileStream.h"
+#include "Core/FileIO/PathUtils.h"
+#include "Core/Strings/AStackString.h"
+
+// REFLECTION
+//------------------------------------------------------------------------------
+REFLECT_NODE_BEGIN( DependencyListNode, Node, MetaName("Dest") + MetaFile() )
+    REFLECT(        m_Source,                   "Source",                   MetaFile() )
+    REFLECT(        m_Dest,                     "Dest",                     MetaFile() )
+    REFLECT_ARRAY(  m_Patterns,                 "Patterns",                 MetaOptional() )
+    REFLECT_ARRAY(  m_PreBuildDependencyNames,  "PreBuildDependencies",     MetaOptional() + MetaFile() + MetaAllowNonFile() )
+REFLECT_END( DependencyListNode )
+
+// AddDependenciesToList()
+//------------------------------------------------------------------------------
+static void AddDependenciesToList( Array< const AString * > * list , const Array< AString > & patterns , const Dependencies & deps )
```

Comment:
This function doesn't seem right.

It's called twice, once with m_StaticDependencies and once with m_DynamicDependencies.

As noted lower down m_DynamicDependencies is always empty, so this call is not needed.

For the first call, with m_StaticDependencies, this will iterate the loop once and compare that name of the one thing you depend on against the patterns. Instead, we should get the dependencies of that object and iterate those.

(In the previous iteration, you were recursively iterating everything, which we don't want, but you do need to go down one level)